app/models/user.rb

```diff
@@ -74,4 +74,68 @@ def update_profile(params)
     def refresh_secret
         self.secret = SecureRandom.base64(24)
     end
+
+    def matches
+        all_users = User.all
+        sorted_list = []
+        for other_user in all_users
+            if other_user.id != self.id
+                match_data = self.score_user(other_user)
+                sorted_list.push({:user => other_user, :match_data => match_data})
+            end
+        end
+        return sorted_list.sort { |e1, e2| e1[:match_data][:score] <=> e2[:match_data][:score] } .reverse.last(5)
+    end
+
+    def score_user(other_user)
+        max = 0
+        max_topic = nil
+
+        common_topics = self.find_common_topics(other_user)
+        for topic in common_topics
+            topic_score = self.find_topic_score(other_user, topic)
+            if max < topic_score
+                max = topic_score
+                max_topic = topic
+            end
+        end
+        return {:score => max, :topic => max_topic }
```

Comment:
A general comment on matching: Why not just give all those users with no common topics zero scores and leave them in the list of possible matches anyway? This way we can just continue down the list to propose possible match candidates without having to write special cases. We are assuming there are lots of users so no having common topic shouldn't be a problem.